internal/service/gateway/tunnel.go

```diff
@@ -0,0 +1,496 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/UpCloudLtd/terraform-provider-upcloud/internal/utils"
+	"github.com/UpCloudLtd/upcloud-go-api/v8/upcloud"
+	"github.com/UpCloudLtd/upcloud-go-api/v8/upcloud/request"
+	"github.com/UpCloudLtd/upcloud-go-api/v8/upcloud/service"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+var allowedSecurityAlgorithms = []string{
+	string(upcloud.GatewayIPSecAlgorithm_aes128),
+	string(upcloud.GatewayIPSecAlgorithm_aes192),
+	string(upcloud.GatewayIPSecAlgorithm_aes256),
+	string(upcloud.GatewayIPSecAlgorithm_aes128gcm16),
+	string(upcloud.GatewayIPSecAlgorithm_aes128gcm128),
+	string(upcloud.GatewayIPSecAlgorithm_aes192gcm16),
+	string(upcloud.GatewayIPSecAlgorithm_aes192gcm128),
+	string(upcloud.GatewayIPSecAlgorithm_aes256gcm16),
+	string(upcloud.GatewayIPSecAlgorithm_aes256gcm128),
+}
+
+var allowedIntegrityAlgorithms = []string{
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_aes128gmac),
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_aes256gmac),
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_sha1),
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_sha256),
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_sha384),
+	string(upcloud.GatewayIPSecIntegrityAlgorithm_sha512),
+}
+
+var allowedDHGroups = []int{2, 5, 14, 15, 16, 18, 19, 20, 21, 24}
+
+func ResourceTunnel() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceTunnelCreate,
+		ReadContext:   resourceTunnelRead,
+		UpdateContext: resourceTunnelUpdate,
+		DeleteContext: resourceTunnelDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description:      "The name of the tunnel, should be unique within the connection",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validateName,
+			},
+			"connection_id": {
+				Description: "ID of the upcloud_gateway_connection resource to which the tunnel belongs",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"local_address_name": {
+				Description:      "Public (UpCloud) endpoint address of this tunnel",
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validateName,
+			},
+			"remote_address": {
+				Description: "Remote public IP address of the tunnel",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"operational_state": {
+				Description: "Tunnel's current operational, effective state",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"ipsec_auth_psk": {
```

Comment:
This is an artificial field in a sense that in API, it is actually nested in `ipsec` block. For example, a POST request to create a tunnel would look like this:
```
{
  "name": "example-tunnel-2",
  "local_address": {
    "name": "public-ip-1"
  },
  "remote_address": {
    "address": "100.10.0.111"
  },
  "ipsec": {
    "authentication": {
        "authentication": "psk",
        "psk": "{your PSK}"
    }
  }
}

```
I have extracted the authentication part to a separate schema block to avoid weird validation issues when we start adding other authentication types. If we need to support a new auth type, we can just implement it as a separate block with it's own set of properties and use `ExactlyOneOf` schema behaviour to make sure user defines only one. This should give us a little more flexibility in the future (I hope :P)